### PR TITLE
add MCAP attachment support for C++

### PR DESCRIPTION
**Pull request by eloff**
### Changelog

Add C++ support for MCAP attachment writing.

### Docs

None

### Description

Builds on the Rust MCAP attachment support added in https://github.com/foxglove/foxglove-sdk/pull/791

Wraps it into C++, following the lead of how we did it for metadata, and adds tests for it.

---

> [!NOTE]
> Adds MCAP attachment writing via new C API and Rust impl, with C++ wrapper method and tests.
> 
> - **C API (`c/include/foxglove-c/foxglove-c.h`)**:
>   - Add `struct foxglove_mcap_attachment` and `foxglove_mcap_attach` to write attachments (non-WASM).
>   - Update cbindgen mappings (`c/cbindgen*.toml`).
> - **Rust FFI (`c/src/channel.rs`)**:
>   - Implement `FoxgloveMcapAttachment` and `foxglove_mcap_attach`, including UTF-8 validation and data handling.
>   - Extend `McapWriterVariant` with `attach` passthrough.
> - **C++ SDK (`cpp/foxglove/include/foxglove/mcap.hpp`, `cpp/foxglove/src/mcap.cpp`)**:
>   - Introduce `foxglove::Attachment` and `McapWriter::attach(const Attachment&)` calling C API.
>   - Declare C bindings for `foxglove_mcap_attach`.
> - **Tests (`cpp/foxglove/tests/test_mcap.cpp`)**:
>   - Add tests for single/multiple/empty attachments and basic verification in written MCAP.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 07be51d04f8e4d061d0fa59cc56ed85bb437cfdd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: cpp/foxglove/include/foxglove/mcap.hpp [hunk 1]
```diff
@@ -64,6 +69,27 @@ enum class McapCompression : uint8_t {
...
+  uint64_t log_time = 0;
+  /// @brief Timestamp at which the attachment was created, in nanoseconds since epoch.
+  /// If not available, set to 0.
+  uint64_t create_time = 0;
+  /// @brief Name of the attachment, e.g. "config.json".
+  std::string_view name;
+  /// @brief Media type of the attachment, e.g. "application/json".
+  std::string_view media_type;
+  /// @brief Pointer to the attachment data.
+  const std::byte* data = nullptr;
```
Comment (clalancette) [inline cpp/foxglove/include/foxglove/mcap.hpp]: This is `const uint8_t *` basically everywhere else.  If we made it a `const uint8_t *` here, we could get rid of the `reinterpret_cast` below.  Any particular reason to use `const std::byte *` here?

File: c/src/channel.rs [hunk 2]
```diff
@@ -410,6 +439,69 @@ unsafe fn do_foxglove_mcap_write_metadata(
...
+
+    let Some(writer_handle) = writer.0.as_ref() else {
+        return Err(foxglove::FoxgloveError::SinkClosed);
+    };
+
+    let data = if attachment.data.is_null() || attachment.data_len == 0 {
+        &[]
+    } else {
+        unsafe { std::slice::from_raw_parts(attachment.data, attachment.data_len) }
+    };
```
Comment (clalancette) [inline c/src/channel.rs:501]: This is minor, but what if `attachment.data != NULL` and `attachment.data_len == 0`, or `attachment.data == NULL` and `attachment.data_len != 0` ?  In both cases, no data will be copied (so we'll avoid UB), but they are both nonsensical operations.  Should we explicitly return an error for these cases?

File: cpp/foxglove/include/foxglove/mcap.hpp (hunk 1 above)
Comment (eloff) [reply cpp/foxglove/include/foxglove/mcap.hpp]: Good point

File: c/src/channel.rs (hunk 2 above)
Comment (eloff) [reply c/src/channel.rs:501]: I think elsewhere we follow this behavior instead of returning an error, but it's an interesting point.

We could log an error, or do that and return an error code, or keep the current behavior of ignoring it. My preference is treat it as an error, but I don't feel strongly about it.

Comment (clalancette) [reply c/src/channel.rs:501]: I see what you mean about following this behavior elsewhere.

I don't feel strongly about it either way.  I also prefer to treat this as an error (much easier to debug).  If you do want to make this an error, I think a way to avoid the Cursor error is to do something like this:

```
let data = if attachment.data.is_null() {
    if attachment.data_len != 0 {
        return Err(foxglove::FoxgloveError::ValueError(
            "attachment data is null but data_len is not 0".to_string(),
        ));
    }
    &[]
} else {
    if attachment.data_len == 0 {
        return Err(foxglove::FoxgloveError::ValueError(
            "attachment data is not null but data_len is 0".to_string(),
        ));
    }
    unsafe { std::slice::from_raw_parts(attachment.data, attachment.data_len) }
};
```

File: cpp/foxglove/include/foxglove/mcap.hpp (hunk 1 above)
Comment (eloff) [reply cpp/foxglove/include/foxglove/mcap.hpp]: Actually it seems there's a lot (more) std::byte usage too. 

std::byte is a little more opaque (requires casts to perform arithmetic) but that doesn't much matter one way or another. uint8_t is problematic sometimes because it's UB to cast uint8_t to an arbitrary type. std::byte, unsigned char are allowed to be cast like that.

I'm tempted to change this back to std::byte fit the other code better.

Slack thread for context: https://foxglove.slack.com/archives/C04R1TNP2TB/p1756332887723899

Comment (clalancette) [reply cpp/foxglove/include/foxglove/mcap.hpp]: For what it is worth, I don't see anything in that thread that suggests it is UB to cast uint8_t to arbitrary types.  In point of fact, the linked [Chromium thread](https://groups.google.com/a/chromium.org/g/cxx/c/bBY0gZa1Otk) suggests the opposite.  Also, the [Stackoverflow thread](https://stackoverflow.com/questions/77097673/a-byte-type-stdbyte-vs-stduint8-t-vs-unsigned-char-vs-char-vs-stdbitset8) states that typeof(uint8_t) === typeof(std::byte) === unsigned char for all common platforms.  The only downside to uint8_t is that it does not necessarily exist on all platforms, but I don't think that applies to anywhere this would run.

All of that said, I really don't feel strongly about it.  My preference is always to reduce casts, and the best way in my experience to reduce casts is to make sure the types are consistent all the way through the stack.  But if we have a preference for using `std::byte`, then we can add that back with casts.

**Review by clalancette — APPROVED** [review]
I'm going to approve this, as I don't really care which way we go with respect to `uint8_t` vs. `std::byte`.

File: cpp/foxglove/include/foxglove/mcap.hpp (hunk 1 above)
Comment (eloff) [reply cpp/foxglove/include/foxglove/mcap.hpp]: I read differently from the C++ thread, see: https://groups.google.com/a/chromium.org/g/cxx/c/bBY0gZa1Otk/m/IE7i5xtqAAAJ and https://groups.google.com/a/chromium.org/g/cxx/c/bBY0gZa1Otk/m/NXgxRpJsAAAJ.

Since we support C++ 17 as the minimum, we can choose std::byte here. I share your preference for keeping it simpler and avoiding casts, and I think in the real world uint8_t will just work. But it does seem to be more problematic with respect to UB and the language spec. Even though it's certainly all unsigned char on any platform we care about.

Given we already seem to prefer std::byte (and changing all that to uint8_t is a breaking change) I think we should just double down on it rather than mix types or try to convert it all to uint8_t.
